### DISCUSSION TITLE
how to use x-collapse in mobile view only?

### DISCUSSION THREAD

**Discussion by StefanVesper** [alpinejs/alpine#2282] [1. Help] [answered]
Hey there! 

First of all, I´d like to thank for AlpineJS, I keep falling in love the longer I get into it! Great work! 

Right now I am trying to find a way to use "x-collapse" only on specific viewport sizes, eg. <1024px.
 
My first idea was to use a function on init of the component that reads the viewportWidth and then adds/removes the "x-collapse" property to/from the $el, but this only works once, on page load. 

 `Alpine.data('flyout', () => ({     
  init() {    
      if(window.innerWidth < 1024) {
          this.$el.setAttribute("x-collapse","")
      }   
  },`

So I bind another function "check" to "@resize.window" and the property gets added and removed from the element succesfully, but after it was added once, the removal doesn´t have any effects on the component, it still uses x-collapse.  

`Alpine.data('flyout', () => ({     
    init() {    
        if(window.innerWidth < 1024) {
            this.$el.setAttribute("x-collapse","")
        }   
    },
    check() {
        if(window.innerWidth > 1024) {
            if(this.$el.hasAttribute("x-collapse")) {
                this.$el.removeAttribute("x-collapse")
            }          
        } else {
            if(!this.$el.hasAttribute("x-collapse")) this.$el.setAttribute("x-collapse","")
        }
    }
}))`

Is there a way to "rerender" the component after removing the property from it?

**Comment by SimoTod** [comment, marked as answer]
It's a weird requirement, I must admit, however, you shouldn't need to do complicate stuff. 
The issue is that collapse doesn't implement a proper clean up face as it could do.
Basically, it just misses
```javascript
        cleanup(() => {
            delete el._x_transition
            el.style.removeProperty('overflow')
            el.style.removeProperty('height')
            if (! el._x_isShown) el.style.display = 'none'
        })
```
As a temporary workaround, you can do something like `delete this.$el._x_transition; this.$el.style.removeProperty('overflow'); this.$el.style.removeProperty('height'); if (! this.$el._x_isShown) this.$el.style.display = 'none'` after `this.$el.removeAttribute("x-collapse")`

Also, you should debounce `@resize.window` otherwise it fires hundreds of times at every resize.

**Reply by StefanVesper** [reply]
works like charm, thank you!

**Reply by StefanVesper** [reply]
I guess since the last update some weird things happen here. 

When I set the "x-collapse" property on window.resize, the word "hidden" gets added as a property to the element when x-collapse is removed , resulting in the element being invisible. the CSS Console sais [hidden] {display:none}  

Could this be some sort of cleanup bug? 

![image](https://user-images.githubusercontent.com/82936840/144076147-c1e552d1-4c2a-4888-8bb2-11c6c69f05f2.png)

**Reply by philwolstenholme** [reply]
Hi @StefanVesper - the `hidden` attribute is related to a new feature I added to Alpine in https://github.com/alpinejs/alpine/pull/2353

One way to remove the attribute would be to do this alongside your other cleanup bits:

```
this.$el.removeAttribute('hidden');
```

**Comment by SimoTod** [comment]
You probably don't want to remove it because it fixes an accessibility
issue. If its to keep borders and margins, you can use a wrapping div and
move the styles you need to show when collapsed to that element

**Reply by StefanVesper** [reply]
I am also removing the hidden attribute now after removing x-collapse. When I remove x-collapse because window.innerWidth is over 1024px, the hidden attribute is no longer needed, right? So it should be fine to remove it in this case without breaking accessibility, correct?

**Reply by SimoTod** [reply]
Yeah, the cleanup function must remove any attribute added by x-collapse

**Comment by timogwl** [comment]
Got the exact same problem. And I don't think it's a weird requirement.

I have a main navigation with dropdowns for each main nav item. On desktop I want to show the dropdown on `@mouseenter` / `@mouseleave` with a `x-transition` and on mobile I want to show the dropdown with `@click` on a button and the `x-collapse` effect.

So I figured out, that it's possible to conditionally bind the transition like this:

```
<div
  x-show="open"
  x-data="{
    transition: {
      'x-transition:enter': 'transition ease-out duration-200 transform',
      'x-transition:enter-start': 'opacity-0 -translate-y-2',
      'x-transition:enter-end': 'opacity-100 translate-y-0',
      'x-transition:leave': 'transition ease-out duration-200',
      'x-transition:leave-start': 'opacity-100',
      'x-transition:leave-end': 'opacity-0'
    }
  }"
  x-bind="$store.nav.isDesktop ? transition : ''"
  x-cloak
>
```

But I didn't find a way to conditionally bind `x-collapse` only for mobile. Is there any way to achieve this without hacky removing attributes and properties?

**Reply by ekwoka** [reply]
Even easier.

On larger viewports just apply `display: contents`.

[this site](https://www.osoandme.com/) uses this pattern in the footer accordion..

[here is a demo](https://awesomealpine.com/play/?coreplugins=2&tw=1&script=Y29uc29sZS5sb2coJ2hlbGxvJyk7Cg&html=PGRpdgogIHgtZGF0YT0ie29wZW46ZmFsc2V9IgogIGNsYXNzPSJmbGV4IGp1c3RpZnktY2VudGVyIGl0ZW1zLWNlbnRlciB3LWZ1bGwgaC1zY3JlZW4iCj4KICA8ZGl2IGNsYXNzPSJmbGV4IGZsZXgtY29sIGdhcC0yIj4KICAgIDxidXR0b24KICAgICAgY2xhc3M9InNtOmhpZGRlbiBweS0yIHB4LTQgYmctd2hpdGUgcm91bmRlZCIKICAgICAgQGNsaWNrPSJvcGVuXj10cnVlIgogICAgPgogICAgICBUb2dnbGUgQ29sbGFwc2UKICAgIDwvYnV0dG9uPgogICAgPGRpdgogICAgICB4LWNvbGxhcHNlCiAgICAgIHgtc2hvdz0ib3BlbiIKICAgICAgY2xhc3M9InNtOiFjb250ZW50cyBiZy13aGl0ZSBzaGFkb3cgcm91bmRlZCIKICAgID4KICAgICAgPGRpdgogICAgICAgIGNsYXNzPSJiZy13aGl0ZSBtYXgtc206c2hhZG93IG1heC1zbTpyb3VuZGVkIGZsZXggZmxleC1jb2wgZ2FwLTIgcC0yIgogICAgICA-CiAgICAgICAgPHNwYW4-IGhlbGxvIHdvcmxkPC9zcGFuPgogICAgICAgIDxzcGFuPiBoZWxsbyB3b3JsZDwvc3Bhbj4KICAgICAgICA8c3Bhbj4gaGVsbG8gd29ybGQ8L3NwYW4-CiAgICAgICAgPHNwYW4-IGhlbGxvIHdvcmxkPC9zcGFuPgogICAgICAgIDxzcGFuPiBoZWxsbyB3b3JsZDwvc3Bhbj4KICAgICAgICA8c3Bhbj4gaGVsbG8gd29ybGQ8L3NwYW4-CiAgICAgICAgPHNwYW4-IGhlbGxvIHdvcmxkPC9zcGFuPgogICAgICAgIDxzcGFuPiBoZWxsbyB3b3JsZDwvc3Bhbj4KICAgICAgPC9kaXY-CiAgICA8L2Rpdj4KICA8L2Rpdj4KPC9kaXY-Cg)

**Reply by timogwl** [reply]
Hey @ekwoka , thanks for your reply!

Unfortunately that's not what I want to achieve. I have a main (header) navigation. Each nav item has a dropdown, which isn't shown initially. 

On desktop I want to show the dropdown, when you hover over the nav item (`@mouseenter` / `@mouseleave`). The dropdown should be shown with a nice `x-transition`.

On mobile I want to show the dropdown, when you `@click` on a button next to the nav item with the nice `x-collapse` effect.

If I put the `x-collapse` on the dropdown, then the `x-transition` isn't working anymore. So I need to conditionally use `x-transition` only for desktop and `x-collapse` only for mobile.

I hope I explained this clearly enough.

**Reply by ekwoka** [reply]
well, you can bind `x-collapse` in the same way you're binding the transitions.

alternatively, you can just use different transitions. a collapse animation is actually quite easy with modern things.

You could even nest them to be extra wacky.

```
<div collapse> // md:!contents
  <div transitions> // max-md:!contents
```

Beautiful.

**Reply by timogwl** [reply]
Thanks, I'll give it a try later!

Do you have an example for me how to bind `x-collapse` the same way I'm binding the transitions? Couldn't find any examples.

**Reply by ekwoka** [reply]
Sure

```html
x-bind="$store.nav.isDesktop ? transition : { 'x-collapse': '' }"
```

seems like it should do it :)

**Reply by timogwl** [reply]
Thanks, it's working! The only downside is, that it's not reactive. `$store.nav.isDesktop` is updated on window resize, but `x-bind` checks the condition only once. But I think I can live with that.

**Reply by ekwoka** [reply]
:(

The stacked divs would work.
